docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,369 @@
+* Facilitate the easy addition of new `out-of-band` task(s) through this custom resource.
+* Provide CLI capabilities to operators making it easy to invoke supported `out-of-band` tasks.
+
+## Non-Goals
+* In the current scope, capability to abort/cancel an out-of-band task is not going to be provided. This could be considered as an enhancement based on pull.
+* Ordering (by establishing dependency) of out-of-band tasks submitted for the same etcd cluster is not been considered in the first increment. In a future version based on how operator tasks are used we will enhance this proposal and the implementation.
+
+
+## Proposal
+
+Authors propose creation of a new single dedicated custom resource to represent an `out-of-band` task. Druid will be enhanced to process the task requests and update its status which can then be tracked/observed.
+
+### API
+
+`ETCDOperatorTask` is the new custom resource that will be introduced. This API will be in `v1alpha1` version and will be subject to change. We will be respecting [Kubernetes Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/).
+
+```yaml
+apiVersion: druid.gardener.cloud/v1alpha1
+kind: EtcdOperatorTask
+metadata:
+    name: <name of operator task resource>
+    namespace: <cluster namespace>
+    generation: <specific generation of the desired state>
+spec:
+    taskType: <type/category of supported out-of-band task>
+    ttlSecondsAfterFinished: <time-to-live to garbage collect the custom resource>
+    taskConfig: <task specific configuration>
+status:
+    observedGeneration: <specific observedGeneration of the resource>
+    taskStatus: < overall status of the task >
+    initiatedAt: <time of intiation of this operation>
+    lastErrors:
+      - code: <error-code>
+        description: <description of the error>
+        lastUpdateTime: <time the error was reported>
+    lastOperation:
+      name: <operation-name>
+      state: <task state as seen at the completion of last operation>
+      lastTransitionTime: <time of transition to this state>
+      reason: <reason/message if any>
+    conditions:
+      - type: <type of condition>
+        status: <status of the condition, one of True, False, Unknown>
+        lastTransitionTime: <last time the condition transitioned from one status to another>
+        reason: <programmatic identifier indicating the reason for the condition's last transition>
+        message: <human readable message indicating details about the transition>
+```
+
+> NOTE: The above custom resource YAML serves as a template which will be used to further create a CRD and Golang APIs.
+
+### Golang API
+
+```go
+// EtcdOperatorTask represents an out-of-band operator task.
+type EtcdOperatorTask struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  // Spec is the specification of the task.
+  Spec EtcdOperatorTaskSpec
+  // Status is most recently observed status of the task.
+  Status EtcdOperatorTaskStatus
+}
+```
+
+#### Spec
```

Comment:
In the member proposal, we took a different approach. Each section added a specific section to the `EtcdMember.Status` and we went about explaining why that was needed and what needs to be recorded for that functionality. This was required because we try to capture lot of things as part of the `Status`. Also in the member proposal we have not given any golang API. In this proposal the CRD is pretty small and simple so we really does not make much of a difference actually. YAML presents a CRD template and then comes the golang API.